### PR TITLE
Darwin: Refactor MTRAsyncWorkQueue to avoid retain cycles

### PR DESCRIPTION
**Pull request by ksperling-apple**
- Pass a completion to readyHandler instead of capturing the work item
- Hold the context weakly within the queue and handle context loss
- Simplify thread-safety by having clear ownership rules for work items
- Consolidate headers (don't need _internal.h anymore)
- Invalidate work queue from `-[MTRDevice invalidate]`

### CODE REVIEW

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 1]
```diff
@@ -15,325 +15,285 @@
...
+    _runningWorkItemCount = 1;
+
+    MTRAsyncWorkItem * workItem = _items.firstObject;
+
+    // Check if batching is possible or needed. Only ask work item to batch once for simplicity
+    auto batchingHandler = workItem.batchingHandler;
+    if (batchingHandler && workItem.retryCount == 0) {
+        while (_items.count >= 2) {
+            MTRAsyncWorkItem * nextWorkItem = _items[1];
+            if (!nextWorkItem.batchingHandler || nextWorkItem.batchingID != workItem.batchingID) {
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:248]: I think the previous setup sort of considered situations where an item is batchable (so can be merged into previous items) without having a batching handler (i.e. can't have later things merged into it).

Looks like this was not used in practice, so this change to just checking batchingHandler here is probably ok....

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h [hunk 2]
```diff
@@ -19,66 +19,177 @@
...
+
+/// Called by the work queue to cancel the work item.
+///
+/// Will be called on the dispatch queue associated with this item.
+/// The work item may or may not have been started already.
+@property (nonatomic, strong, nullable) void (^cancelHandler)(void);
+
+@property (nonatomic, readonly) NSUInteger batchingID;
+@property (nonatomic, readonly) id batchableData;
+@property (nonatomic, readonly) MTRAsyncWorkBatchingHandler batchingHandler;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h:124]: Aren't these nullable?  Certainly we are nil-checking batchingHandler.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h [hunk 3]
```diff
@@ -19,66 +19,177 @@
...
+/// Note: This handler is NOT called on the dispatch queue associated with
+/// this work item. Thread-safety is managed by the work queue internally.
+///
+/// @see MTRAsyncWorkBatchingHandler
+- (void)setBatchingID:(NSUInteger)opaqueBatchingID
+                 data:(id)opaqueBatchableData
+              handler:(MTRAsyncWorkBatchingHandler)batchingHandler;
+
+@property (nonatomic, readonly) NSUInteger duplicateTypeID;
+@property (nonatomic, readonly) MTRAsyncWorkDuplicateCheckHandler duplicateCheckHandler;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h:137]: Again, this is nullable.

We should also document which properties should be ignored when other properties are nil....

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 4]
```diff
@@ -15,325 +15,285 @@
...
+        NSAssert(NO, @"work item to post-process is not running");
+        return;
     }
-    os_unfair_lock_unlock(&_lock);
-}
 
-- (void)endWork
-{
-    [self.workQueue endWork:self];
-    [self invalidate];
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:278]: This is hard to parse because the whole file got reordered so the diff is not very readable, but I believe this used to drop references to the handlers, and we are no longer doing that.  

The dropping references was a cycle-breaking measure, but it still seems like a good idea, no?  As things stand we could have handlers that accidentally close over the work item and then we will leak.

Concretely, at the points when we mark a work item complete we should drop all its refs to handlers.

File: src/darwin/Framework/CHIP/MTRDevice.mm [hunk 5]
```diff
@@ -882,20 +885,18 @@ static BOOL AttributeHasChangesOmittedQuality(MTRAttributePath * attributePath)
...
         NSMutableArray<NSArray *> * readRequests = [NSMutableArray arrayWithObject:readRequestData];
 
         // Create work item, set ready handler to perform task, then enqueue the work
         MTRAsyncWorkItem * workItem = [[MTRAsyncWorkItem alloc] initWithQueue:self.queue];
-        MTRAsyncWorkBatchingHandler batchingHandler = ^(id opaqueDataCurrent, id opaqueDataNext, BOOL * fullyMerged) {
+        [workItem setBatchingID:MTRDeviceWorkItemBatchingReadID data:readRequests handler:^(id opaqueDataCurrent, id opaqueDataNext, BOOL * fullyMerged) {
             NSMutableArray<NSArray *> * readRequestsCurrent = opaqueDataCurrent;
             NSMutableArray<NSArray *> * readRequestsNext = opaqueDataNext;
 
-            *fullyMerged = NO;
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:897]: It’s pretty standard to have output parameters like `stop` etc initialised by the caller to default values. I think relying on every block instance to do this is more problematic. I would probably prefer if this was a return value instead of an out parameter…

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 6]
```diff
@@ -15,325 +15,285 @@
...
+        [item cancel];
     }
+    [_items removeAllObjects];
     os_unfair_lock_unlock(&_lock);
 }
 
-- (void)setCancelHandler:(dispatch_block_t)cancelHandler
+- (void)_postProcessWorkItem:(MTRAsyncWorkItem *)workItem retry:(BOOL)retry
 {
-    os_unfair_lock_lock(&_lock);
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:268]: The lock is now taken by the caller. I’ve moved the lock calls up into the public methods in all cases now (and have them more obviously balanced) and the helper methods just assume it is held.

File: src/darwin/Framework/CHIP/MTRDevice.mm (hunk 5 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:897]: I am not sure why it wasn't made a return value to start with, though @jtung-apple might recall.  Maybe a followup to change that?

In any case, I am OK with the caller pre-initializing to NO, but needs to be documented.

[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:897]: I didn't make these return values because the meaning of the return value wasn't clear from the name of these handlers. At the time I felt that this is the clearest way to have code self-document.

I'm not opposed to having it as return value, but I think it needs to be documented well. Either have the handler names be named sufficiently well to imply the meaning of the return, or have it return an enum, which can be clearer than this.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 6 above)
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:268]: I second adding a `os_unfair_lock_assert_owner`.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h [hunk 7]
```diff
@@ -19,66 +19,177 @@
...
+/// least a `readyHandler` (though it is technically optional).
+///
+/// This class is not thread-safe, and once a work item has be submitted to the
+/// queue via `enqueueWorkItem` ownership of the work item passes to the queue.
+/// No further modifications may be made to it after that point.
+///
+/// @see -[MTRAsyncWorkQueue enqueueWorkItem:]
 MTR_TESTABLE
-@interface MTRAsyncWorkQueue : NSObject
+@interface MTRAsyncWorkItem<__contravariant ContextType> : NSObject
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h:93]: No, it means if you have context types `BaseC` and `SubC : BaseC` then a `WorkItem<BaseC>` is a sub-type of a `WorkItem<SubC>`, so the `WorkItem<BaseC>` can be submitted to a `WorkQueue<SubC>`.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 8]
```diff
@@ -15,325 +15,285 @@
...
-        [item cancel];
-    }
-    [invalidateItems removeAllObjects];
+    [self assertMutable];
+    _cancelHandler = cancelHandler;
 }
 
-// called after executing a work item
-- (void)_postProcessWorkItem:(MTRAsyncWorkItem *)workItem retry:(BOOL)retry
+- (void)markedEnqueued
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:65]: Somehow I managed to read that as "markEnqueued" all along :-)

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 9]
```diff
@@ -15,325 +15,285 @@
...
-                }
-
-                [self.items removeObjectAtIndex:1];
-            }
+    if (_state != MTRAsyncWorkItemComplete) {
+        _state = MTRAsyncWorkItemComplete;
+        auto cancelHandler = _cancelHandler;
+        if (cancelHandler) {
+            // Note that this does not prevent a race against
+            // the readyHandler calling the work completion.
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:126]: So what (I think) I have implemented is that once a WorkItem has been enqueued (i.e. is owned by the queue) all state changes happen under the queue lock. So there is no race between `_callNextReadyWorkItem` and `invalidate` (and `invalidate` is the only way to cancel a work item currently). Cancellation can of course happen after the readyHandler has already been called (or dispatched to be called), but that's existing and expected behavior to cancel (if possible) a job that's already executing.

The race is due to the possible delay between the queue deciding to cancel a work item and the cancelHandler call actually being run on the dispatch queue. It's possible for the work completion handler to be called in-between these two points in time -- though it will return `NO` because from the perspective of the queue the work is already cancelled, so it's possible for the work code to deal with this.

This race also exists in the existing code; the decision to call the cancelHandler is made by code under the queue and work item locks, and then the dispatch is performed. However this is all academic, because no existing work we have even registers a cancel handler.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 10]
```diff
@@ -15,325 +15,285 @@
...
 
-- (void)setReadyHandler:(MTRAsyncWorkReadyHandler)readyHandler
+- (void)invalidate
 {
     os_unfair_lock_lock(&_lock);
-    if (!_enqueued) {
-        _readyHandler = readyHandler;
+    MTR_LOG_INFO("MTRAsyncWorkQueue<%@> invalidate %tu items", _context, _items.count);
+    for (MTRAsyncWorkItem * item in _items) {
+        [item cancel];
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:194]: As noted above, all state changes to enqueued work items, including `callReadyHandlerWithContext` and `cancel` are made under the queue lock. I'll add a comment to point this out explicitly.

In terms of logging... it's a trade-off between minimizing the work done under the lock, vs actually having the order of log messages reflect the serialized order of things that were going on. Logging outside of the lock has the potential to mess that up. I think in this case, I wouldn't expect this to be a highly contended lock, and know that the context is an MTRDevice in practice, so it's not exactly "arbitrary" code. I guess if we wanted to be pedantic we could call `context.description` before we take the lock and then log it under the lock.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 1 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:248]: `setBatchingID:data:handler:` (and `setDuplicateTypeID:handler`) both had non-null handler parameters, so I wasn't aware of this idea. We can re-add that if needed I guess (or maybe it's more obvious in those cases to implement a handler that just doesn't batch anything).

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h (hunk 3 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h:137]: Ah yes, good catch on the nullable.

In terms of documenting the properties, having these exposed as readable properties at all is somewhat superfluous, since the client code is really just meant to set the handlers it wants and then enqueue the work item. There's no other API exposed on the items that the client itself is meant to use. I guess I can add a line to the `set...` methods to say that if it's not called the item doesn't participate in batching / duplicate checking.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 4 above)
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:278]: I can add that back in... I really dislike this "generic object with a bunch of blocks" instead of "sub-classing and overriding the relevant methods" pattern, because it produces all these sorts of issues. If only ObjC had anonymous sub-classes...

File: src/darwin/Framework/CHIP/MTRDevice.mm (hunk 5 above)
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRDevice.mm:897]: I'll add some explicit comments, but I think it's fairly obvious from when the comment says "to indicate such and such set `*foo` to YES" that it starts out as NO. See the `stop` parameter in https://developer.apple.com/documentation/foundation/nsarray/1415846-enumerateobjectsusingblock for an example.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm [hunk 11]
```diff
@@ -15,325 +15,285 @@
...
-@end
+#import <os/lock.h>
 
-#pragma mark - Class implementations
+typedef NS_ENUM(NSInteger, MTRAsyncWorkItemState) {
+    MTRAsyncWorkItemMutable,
+    MTRAsyncWorkItemComplete,
+    MTRAsyncWorkItemEnqueued,
+    MTRAsyncWorkItemRunning,
+    // values > MTRAsyncWorkItemRunning encode retryCount
```
[the remark this answers is not in the document]
Comment (ksperling-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:29]: That's effectively what `MTRAsyncWorkItemRunning` is because retryCount starts at 0?

[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:29]: Yes functionally the same, but the idea here is for code to document itself - the suggestion is having a more explicit name for an enum (that is the same value as `MTRAsyncWorkItemRunning`) that you use as the base for counting retries.

In other words, when a future engineer looks at the code, `MTRAsyncWorkItemRetryCountBase + 1` makes more sense than  `MTRAsyncWorkItemRunning + 1`.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h:95]: Ah, that makes sense.  Thanks!

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 9 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: > The race is due to the possible delay between the queue deciding to cancel a work item and the cancelHandler call actually being run on the dispatch queue. 

Right.

> This race also exists in the existing code

That's what I said above, yes.  It had a smaller window, since you have to race against linear code, not async-dispatched code, but it was still there.

So the thing is: having a work item do any of its "ready" work after invalidation is pretty dangerous.  It can lead to things being queued on the now-quiescent Matter work queue, where they lie in wait for the next startup.

Looking at how this can be reached (via `invalidate` only, as you note), it looks like it's called on an arbitrary thread/queue (wherever the SDK consumer decides to call `shutdown` on a controller).  

The real invariant we need to preserve is that no ready handler runs after invalidate returns.

It looks like `callReadyHandlerWithContext` and `_callNextReadyWorkItem` are both called with the lock held, right?  I guess that's pre-existing, though it does mean we are calling out into the consumer-provided batching handler with the lock held, which gives me the willies.  But if we want to make sure the blocks callReadyHandlerWithContext submits never run after invalidate, what are our options?

We could have that block close over a weak-self, stongify, take the lock and check whether we have been invalidated or something.  But that would not help the situation when we are in the middle of running a readyHandler already when the invalidate comes in.

Maybe we need a followup for this, because this is sounding hard.  But I suspect we will be getting random crashes for a while if we don't actually end up with something where stack shutdown effectively stops all ongoing work and joins the threads....

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 10 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:208]: > I guess if we wanted to be pedantic we could call context.description before we take the lock and then log it under the lock.

Yes, please!

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 1 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm:265]: I think the new code is fine.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.h]: I think the key part is saying "don't look at duplicateTypeID if duplicateCheckHandler is nil" or so.

File: src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm (hunk 9 above)
[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: If we can assume invalidate is called on the same queue as the work items, then the async could be constructed as a `dispatch_block_create` that is set on the work item. The invalidate call would call `dispatch_block_cancel` on it, and we'd be done.

If we cannot assume invalidate is called, then we may need the above plus something `invalidate` can wait on, if we need nothing to be running after invalidate returns.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: Sadly, the invalidate is called on the framework consumer's queue (when they shut down the controller), but the items in practice run on `"org.csa-iot.matter.framework.device.workqueue"`.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: Note that we can also shut down the device's work queue when the device is invalidated, but that does not help if an item is running on that queue right at that second....

On the other hand, maybe this is a non-issue because almost all of our dispatch to the Matter queue should be going via the controller, and that is getting shut down...  So once it gets far enough we can't dispatch on that queue anymore.  The exceptions are in `MTRBaseSubscriptionCallback`

[the remark this answers is not in the document]
Comment (jtung-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: Actually, I just tested this and we could do `dispatch_block_cancel` and then `dispatch_block_wait` if the invalidate comes from a different queue, to wait until it finishes executing, if the block already started running.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: And I guess the operational browser is an exception too.  But I really hope that `DNSServiceRefDeallocate` will prevent the blocks from the DNS stuff from running after that, and we do that on the Matter queue, so it should be safe.

[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/darwin/Framework/CHIP/MTRAsyncWorkQueue.mm]: I think I have convinced myself this is safe now.

**Review by jtung-apple — APPROVED** [review]

**Comment by ksperling-apple** [comment]
Fast-tracking Darwin-only change.
